pkg/ingester/ingester_v2.go

```diff
@@ -1216,8 +1217,30 @@ func (i *Ingester) compactionLoop(ctx context.Context) error {
 	return nil
 }
 
+// compactHead compact the Head block at 2h intervals,
+// avoiding a single huge block.
+func compactHead(userDB *userTSDB) error {
+	// The Compact() will produces multiple 2h blocks if Head has a lot of data.
+	if err := userDB.Compact(); err != nil {
+		return err
+	}
+	twentyFourHrs := 24 * time.Hour.Milliseconds()
+	h := userDB.Head()
+	if (h.MinTime()/twentyFourHrs)*twentyFourHrs != (h.MaxTime()/twentyFourHrs)*twentyFourHrs {
+		// Remaining data in Head spans across 2 days, so we break it into 2 blocks.
+		maxt := (h.MaxTime()/twentyFourHrs)*twentyFourHrs - 1
+		if err := userDB.CompactHead(tsdb.NewRangeHead(h, h.MinTime(), maxt)); err != nil {
+			return err
+		}
+	}
+	return userDB.CompactHead(tsdb.NewRangeHead(h, h.MinTime(), h.MaxTime()))
+}
+
 // Compacts all compactable blocks. Force flag will force compaction even if head is not compactable yet.
 func (i *Ingester) compactBlocks(ctx context.Context, force bool) {
+	i.TSDBState.compactBlocksMtx.Lock()
+	defer i.TSDBState.compactBlocksMtx.Unlock()
```

Comment:
As the access to the mint and maxt of Head and then compaction is not atomic, we need this protection

Reply:
I would make it per TSDB, what do you think?

If yes, then we need to wrap the call to `userDB.Compact()` to take the lock there too. Moving in this direction, to make the code more clear, I would then embed `*tsdb.DB` with a name in `type userTSDB struct` so that we explicitly wrap the `tsdb.DB` functions we need to call.

Reply:
Makes sense to make it per TSDB

Reply:
Taking a look now, and I think we don't need to make `*tsdb.DB` with a name right now. We don't need a wrapper around `Compact()`, but rather `CompactHead`, but the wrapper wont match the original `CompactHead` function signature. So I am thinking it to keep it simple and have `userDB.compactHead()`. WDYT? Embeding `*tsdb.DB` with a name sounds fine too if that would help later.

Reply:
Oh I see now, it will be easier to use the lock if we also wrap `Compact()`. I will embed the tsdb.DB as a name in the struct.

Reply:
Embedding `tsdb.DB` with a name may help to avoid misuses, because functions must be explicitly wrapped.